platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java

```diff
@@ -266,67 +286,99 @@ public void initInputDevices() {
 		}
 	}
 
+	private int assignJoystickIdNumber(int deviceId) {
+		int godotJoyId = 0;
+		while (mJoystickIds.containsValue(godotJoyId)) {
+			godotJoyId++;
+		}
+		mJoystickIds.put(deviceId, godotJoyId);
+		return godotJoyId;
+	}
+
 	@Override
 	public void onInputDeviceAdded(int deviceId) {
-		int id = findJoystickDevice(deviceId);
-
 		// Check if the device has not been already added
-		if (id < 0) {
-			InputDevice device = mInputManager.getInputDevice(deviceId);
-			//device can be null if deviceId is not found
-			if (device != null) {
-				int sources = device.getSources();
-				if (((sources & InputDevice.SOURCE_GAMEPAD) == InputDevice.SOURCE_GAMEPAD) ||
-						((sources & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK)) {
-					id = mJoysticksDevices.size();
-
-					Joystick joy = new Joystick();
-					joy.device_id = deviceId;
-					joy.name = device.getName();
-					joy.axes = new ArrayList<InputDevice.MotionRange>();
-					joy.hats = new ArrayList<InputDevice.MotionRange>();
-
-					List<InputDevice.MotionRange> ranges = device.getMotionRanges();
-					Collections.sort(ranges, new RangeComparator());
-
-					for (InputDevice.MotionRange range : ranges) {
-						if (range.getAxis() == MotionEvent.AXIS_HAT_X || range.getAxis() == MotionEvent.AXIS_HAT_Y) {
-							joy.hats.add(range);
-						} else {
-							joy.axes.add(range);
-						}
-					}
 
-					mJoysticksDevices.add(joy);
+		if (mJoystickIds.containsKey(deviceId)) {
+			return;
+		}
+
+		InputDevice device = mInputManager.getInputDevice(deviceId);
+		//device can be null if deviceId is not found
+		if (device == null) {
+			return;
+		}
+
+		int sources = device.getSources();
+
+		// Device may not be a joystick or gamepad
+		if (!(((sources & InputDevice.SOURCE_GAMEPAD) == InputDevice.SOURCE_GAMEPAD) ||
+					((sources & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK))) {
+			return;
+		}
```

Comment:
Can you propagate the `!` inside to make it simpler to parse and read.

Reply:
Unless I'm misreading this. I think the following would be better?

```
if ((sources & (InputDevice.SOURCE_GAMEPAD|InputDevice.SOURCE_JOYSTICK)==0) {
    return;
}
```

Reply:
I'm a bit rusty on bit mask operations, but couldn't there be cases where `(sources & InputDevice.SOURCE_GAMEPAD) != InputDevice.SOURCE_GAMEPAD` but is still `!= 0`?

Reply:
"bitwise and" masks out bits not present in the mask.

The op ```(sources & (InputDevice.SOURCE_GAMEPAD|InputDevice.SOURCE_JOYSTICK)``` should result in only four possible values: ```0``` (no matching bits in mask), ```InputDevice.SOURCE_GAMEPAD```, ```InputDevice.SOURCE_JOYSTICK```, or ```(InputDevice.SOURCE_GAMEPAD|InputDevice.SOURCE_JOYSTICK)```.

Anyone else willing to chime in?